astroid/bases.py

```diff
@@ -436,7 +436,14 @@ def _infer_builtin_new(
         node_context = context.extra_context.get(caller.args[0])
         infer = caller.args[0].infer(context=node_context)
 
-        yield from (Instance(x) if x is not Uninferable else x for x in infer)  # type: ignore[misc,arg-type]
+        for inferred in infer:
+            if isinstance(inferred, type(Uninferable)):  # isinstance helps mypy
+                yield inferred
+            if isinstance(inferred, Instance):
+                raise InferenceError
+            if isinstance(inferred, (nodes.ClassDef, nodes.Lambda, Proxy)):
```

Comment:
I thought about this some more, and this (currently) can't be correct as well.

We expect the `proxy` of an `Instance` to be `ClassDef`, see its typing. If this crashes without also allowing `Lambda` or `Proxy` we should fix that typing, if not this should only accept `ClassDef`.

Somewhat related, I think it would be good to add an `__init__` to `Instance` which narrows down the type of `proxied`. Would you be willing to do that in this PR as well?

Reply:
Sure, if we're not backporting that's on the table 👍🏻